{{cookiecutter.project_slug}}/_/ci-services/definitions/.gitlab-ci.yml

```diff
@@ -41,20 +41,32 @@ stages:
                  -f "deployment/application/Dockerfile" .
   - docker push "${IMAGE}"
 
+.generate-secrets:
+  before_script:
+  - kubectl -n ${TARGET} get secret ${APPLICATION_NAME} ||
+    kubectl -n ${TARGET} create secret generic ${APPLICATION_NAME} --from-literal=DJANGO_SECRET_KEY=$(< /dev/urandom tr -dc A-Za-z0-9 | head -c50) |
+    kubectl -n ${TARGET} apply -f -
+  - kubectl -n ${TARGET} get secret ${DATABASE_NAME} ||
+    kubectl -n ${TARGET} create secret generic ${DATABASE_NAME} --from-literal=POSTGRESQL_PASSWORD=$(< /dev/urandom tr -dc A-Za-z0-9 | head -c16) |
+    kubectl -n ${TARGET} apply -f -
+
 .deploy:
   stage: deploy
-  image: docker.io/appuio/oc:v3.9
+  extends: generate-secrets
+  environment:
+    url: "${KUBE_URL}/console/project/{{ cookiecutter.project_slug }}{% if cookiecutter.environment_strategy == 'dedicated' %}-${CI_ENVIRONMENT_NAME}{% endif %}/overview"
+  variables:
+    TARGET: "{{ cookiecutter.project_slug }}{% if cookiecutter.environment_strategy == 'dedicated' %}-${CI_ENVIRONMENT_NAME}{% endif %}"
+    APPLICATION_NAME: "application{% if cookiecutter.environment_strategy == 'shared' %}-${CI_ENVIRONMENT_NAME}{% endif %}"
+    DATABASE_NAME: "database{{ cookiecutter.project_slug }}{% if cookiecutter.environment_strategy == 'shared' %}-${CI_ENVIRONMENT_NAME}{% endif %}"
+  image: docker.io/appuio/oc:v3.11
   script:
-  - oc project ${TARGET}
-  - oc get secret postgres ||
-    oc process -f deployment/postgres-secrets.yaml |
-    oc apply --overwrite -f -
-  - oc get secret application ||
-    oc process -f deployment/application-secrets.yaml |
-    oc apply --overwrite -f -
   - oc tag "${SOURCE}/${CI_PROJECT_NAME}:${CI_COMMIT_SHA}"
            "${TARGET}/${CI_PROJECT_NAME}:${CI_COMMIT_SHA}"
-  - oc process --param-file=deployment/envs/${CI_ENVIRONMENT_NAME} -p APP_TAG=${CI_COMMIT_SHA} -f deployment/postgres.yaml |
-    oc apply --overwrite -f -
-  - oc process --param-file=deployment/envs/${CI_ENVIRONMENT_NAME} -p APP_TAG=${CI_COMMIT_SHA} -f deployment/application.yaml |
-    oc apply --overwrite -f -
+  - pushd deployment/kubernetes/application/base &&
+    kustomize edit set image IMAGE="${DOCKER_REGISTRY}/${TARGET}/${CI_PROJECT_NAME}:${CI_COMMIT_SHA}" &&
+    popd
+  - kustomize build deployment/kubernetes/application/${CI_ENVIRONMENT_NAME} | envsubst | kubeval --strict --ignore-missing-schemas
+  - kustomize build deployment/kubernetes/database/${CI_ENVIRONMENT_NAME} | envsubst | kubeval --strict
+  - kustomize build deployment/kubernetes/application/${CI_ENVIRONMENT_NAME} | envsubst | kubectl -n ${TARGET} apply -f -
+  - kustomize build deployment/kubernetes/database/${CI_ENVIRONMENT_NAME} | envsubst | kubectl -n ${TARGET} apply -f -
```

Comment:
Why do you add the `kubernetes/` directory level? Is this necessary? Let's assume we run an opinionated setup, Kubernetes is the only choice.

Reply:
Because we already have `application` at the deployment level containing the Dockerfile and entrypoint script.